### PR TITLE
enable workspace lints for app/crates

### PR DESCRIPTION
**Pull request by uttkarzsh**
Resolves issue https://github.com/NethermindEth/stellar-private-payments/issues/185 and enables workspace lints for `app/crates`

### CODE REVIEW

File: app/crates/core/state/src/storage.rs [hunk 1]
```diff
@@ -575,6 +575,7 @@ impl Storage {
         };
 
         if current_ledger > sync_meta.last_ledger {
+            #[allow(clippy::arithmetic_side_effects)]
```
Comment (maksimryndin) [inline app/crates/core/state/src/storage.rs]: hey @uttkarzsh ✋ thank you! :) 

enabling lints - let's truly fix the code rather than add `#[allow...]`. E.g. for this case `current_ledger.checked_sub(...`

Comment (uttkarzsh) [reply app/crates/core/state/src/storage.rs]: Updated 😄

File: app/crates/core/state/src/storage.rs [hunk 2]
```diff
@@ -1212,7 +1212,10 @@ mod tests {
 
         let meta = storage.get_sync_metadata()?;
         assert!(meta.is_some());
-        let meta = meta.unwrap();
```
Comment (maksimryndin) [inline app/crates/core/state/src/storage.rs:1215]: could we use just `expect("expected sync metadata")` here?

File: app/crates/core/state/src/storage.rs [hunk 3]
```diff
@@ -1223,7 +1226,10 @@ mod tests {
             events: vec![],
         })?;
 
-        let meta = storage.get_sync_metadata()?.unwrap();
+        let meta = match storage.get_sync_metadata()? {
```
Comment (maksimryndin) [inline app/crates/core/state/src/storage.rs]: ditto

File: app/crates/core/stellar/src/contract_state.rs [hunk 4]
```diff
@@ -399,10 +398,15 @@ impl StateFetcher {
     fn field_to_scval_u256(v: Field) -> xdr::ScVal {
         let be = v.to_be_bytes();
 
-        let hi_hi = u64::from_be_bytes(be[0..8].try_into().unwrap());
```
Comment (maksimryndin) [inline app/crates/core/stellar/src/contract_state.rs:402]: could we keep slices conversion here and just use `expect`? it is a compile-time check

File: app/crates/core/stellar/src/conversions.rs [hunk 5]
```diff
@@ -49,7 +50,11 @@ pub fn scval_to_u256(val: &xdr::ScVal) -> Result<U256, Error> {
         let lo_hi = U256::from(parts.lo_hi);
         let lo_lo = U256::from(parts.lo_lo);
 
-        Ok((hi_hi << 192) + (hi_lo << 128) + (lo_hi << 64) + lo_lo)
+        let mut out = Shl::shl(hi_hi, 192);
```
Comment (maksimryndin) [inline app/crates/core/stellar/src/conversions.rs:53]: why do we need here `Shl`? Could we just

`Ok((hi_hi << 192).checked_add(hi_lo << 128)?.checked_add(lo_hi << 64)?.checked_add(lo_lo)?`

?

File: app/crates/core/stellar/src/ext_data_hash.rs [hunk 6]
```diff
@@ -57,7 +58,7 @@ pub fn hash_ext_data_offchain(ext: &ExtData) -> Result<[u8; 32]> {
     digest_be.copy_from_slice(digest.as_slice());
     let digest_u256 = U256::from_big_endian(&digest_be);
     let modulus = U256::from_big_endian(&BN254_MODULUS_BE);
-    let reduced = digest_u256 % modulus;
+    let reduced = Rem::rem(digest_u256, modulus);
```
Comment (maksimryndin) [inline app/crates/core/stellar/src/ext_data_hash.rs:61]: So was here a lint error? That we wrap into `Rem`?

File: app/crates/core/stellar/src/ext_data_hash.rs [hunk 7]
```diff
@@ -68,10 +69,18 @@ pub fn hash_ext_data_offchain(ext: &ExtData) -> Result<[u8; 32]> {
 /// Correctly maps i128 to Soroban's I256 XDR representation
 fn i128_to_i256_scval(n: i128) -> ScVal {
     let hi = if n < 0 { -1i64 } else { 0i64 };
+    let hi_lo = u64::from_be_bytes(hi.to_be_bytes());
+    let bytes = n.to_be_bytes();
+    let lo_hi = u64::from_be_bytes([
+        bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
```
Comment (maksimryndin) [inline app/crates/core/stellar/src/ext_data_hash.rs]: Could we use slices here with `expect`?

File: app/crates/core/types/src/amounts.rs [hunk 8]
```diff
@@ -456,17 +472,17 @@ impl Add for Field {
 
     fn add(self, rhs: Self) -> Self::Output {
         let m = Field::modulus();
-        let mut v = self.0 + rhs.0;
-        if v >= m {
-            v -= m;
+        let (mut v, overflow) = self.0.overflowing_add(rhs.0);
```
Comment (maksimryndin) [inline app/crates/core/types/src/amounts.rs]: let's add or verify that we have tests for these operations (`Add`, `Sub`) with edge cases?

**Review by maksimryndin — COMMENTED** [review]
Awesome @uttkarzsh ! 👍  thanks, some comments to address

File: app/crates/core/stellar/src/conversions.rs (hunk 5 above)
Comment (uttkarzsh) [reply app/crates/core/stellar/src/conversions.rs:53]: hey @maksimryndin . 👋🏻. clippy is failing with `arithmetic_side_effects` when we use shifts or modulus. That's why i used `Shl` and `Rem`.


Screenshot 2026-04-30 at 03 19 54

Comment (maksimryndin) [reply app/crates/core/stellar/src/conversions.rs:53]: why overflowing add - it may introduce incorrectness - let's use `checked_add`?

**Review by maksimryndin — COMMENTED** [review]
@uttkarzsh all good - just let's put back slices and there is a question on `overflowing_add`

**Comment by uttkarzsh** [comment]
@maksimryndin. i updated the slices with `expect()` and replaced `overflowing_add()` and `overflowing_sub()` with `checked_add()` and `chekced_sub()` . lmk if anything else is needed.

File: app/crates/core/stellar/src/conversions.rs [hunk 9]
```diff
@@ -49,7 +50,11 @@ pub fn scval_to_u256(val: &xdr::ScVal) -> Result<U256, Error> {
...
+        out = out.overflowing_add(Shl::shl(hi_lo, 128)).0;
+        out = out.overflowing_add(Shl::shl(lo_hi, 64)).0;
```
Comment (maksimryndin) [inline app/crates/core/stellar/src/conversions.rs]: @uttkarzsh here the overflowing - can we use checked_add?

Comment (uttkarzsh) [reply app/crates/core/stellar/src/conversions.rs]: fixed :)

**Review by maksimryndin — APPROVED** [review]
amazing @uttkarzsh ! 👍  thanks a lot for the contribution! 🚀
